vajram/extensions/graphql/vajram-graphql-codegen/src/main/java/com/flipkart/krystal/vajram/graphql/codegen/GraphQlCodeGenUtil.java

```diff
@@ -99,37 +99,67 @@ TypeName getTypeNameForField(WrappedType fieldType, GraphQlFieldSpec fieldSpec)
   }
 
   ClassName getTypeNameForField(PlainType fieldType, GraphQlFieldSpec fieldSpec) {
-    GraphQLTypeName typeName = new GraphQLTypeName(fieldType.graphQlType().getName());
-    String packageName = null;
+    String graphQlTypeName = fieldType.graphQlType().getName();
+
+    ClassName fieldLevelOverride = getFieldLevelJavaType(fieldSpec);
+    if (fieldLevelOverride != null) {
+      return fieldLevelOverride;
+    }
+
+    Optional<ClassName> scalarJavaType = schemaReaderUtil.getJavaTypeForScalar(graphQlTypeName);
+
+    return scalarJavaType.orElseGet(
+        () ->
+            switch (graphQlTypeName) {
+              case "String" -> ClassName.get(String.class);
+              case "Int" -> ClassName.get(Integer.class);
+              case "Boolean" -> ClassName.get(Boolean.class);
+              case "Float" -> ClassName.get(Float.class);
+              case "ID" -> {
+                GraphQLTypeName enclosingType = fieldSpec.enclosingType();
+                yield enclosingType != null
+                    ? schemaReaderUtil.entityIdClassName(enclosingType)
+                    : ClassName.get(Object.class);
+              }
+              default -> {
+                GraphQLTypeName typeName = new GraphQLTypeName(graphQlTypeName);
+                yield ClassName.get(
+                    schemaReaderUtil.getPackageNameForType(typeName), typeName.value());
+              }
+            });
+  }
+
+  /**
+   * Extracts the Java type from a field-level @javaType directive if present. This allows per-field
+   * override of the Java type, taking highest priority over scalar-level mappings.
+   */
+  private @Nullable ClassName getFieldLevelJavaType(GraphQlFieldSpec fieldSpec) {
```

Comment:
I think allowing field level override in-general will make the code more complex since same graphql type would be deserialized as different types in different cases - this can lead to hard-to-catch bugs as well. Can we restrict java type to scalar level only?